File: packages/contracts/src/quality.ts
import { z } from "zod";
import { columnTypeSchema, datasetIdSchema } from "./dataset.js";

const columnNameSchema = z.string().min(1).max(500);
export const validationRuleKindSchema = z.enum([
  "required",
  "unique",
  "number-range",
  "pattern",
  "allowed-values",
]);

export const validationRuleSchema = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("required"), column: columnNameSchema }).strict(),
  z.object({ kind: z.literal("unique"), column: columnNameSchema }).strict(),
  z
    .object({
      kind: z.literal("number-range"),
      column: columnNameSchema,
      minimum: z.number().finite().nullable().default(null),
      maximum: z.number().finite().nullable().default(null),
    })
    .strict(),
  z
    .object({
      kind: z.literal("pattern"),
      column: columnNameSchema,
      pattern: z.string().min(1).max(200),
    })
    .strict(),
  z
    .object({
      kind: z.literal("allowed-values"),
      column: columnNameSchema,
      values: z.array(z.string().max(500)).min(1).max(50),
    })
    .strict(),
]);

export const validationRulesSchema = z.array(validationRuleSchema).max(100).superRefine((rules, context) => {
  for (const [index, rule] of rules.entries()) {
    if (rule.kind === "number-range") {
      if (rule.minimum === null && rule.maximum === null) {
        context.addIssue({ code: "custom", message: "A number range needs at least one bound", path: [index] });
      }
      if (rule.minimum !== null && rule.maximum !== null && rule.minimum > rule.maximum) {
        context.addIssue({ code: "custom", message: "The minimum cannot exceed the maximum", path: [index] });
      }
    }
    if (rule.kind === "allowed-values" && new Set(rule.values).size !== rule.values.length) {
      context.addIssue({ code: "custom", message: "Allowed values must be unique", path: [index, "values"] });
    }
  }
});

export const datasetValidationSaveInputSchema = z
  .object({
    datasetId: datasetIdSchema,
    rules: validationRulesSchema,
  })
  .strict();

export const columnQualitySchema = z
  .object({
    name: columnNameSchema,
    inferredType: columnTypeSchema,
    rowCount: z.number().int().nonnegative(),
    nullCount: z.number().int().nonnegative(),
    nullRate: z.number().min(0).max(1),
    distinctCount: z.number().int().nonnegative(),
    distinctRate: z.number().min(0).max(1),
    minValue: z.string().nullable(),
    maxValue: z.string().nullable(),
  })
  .strict();

export const qualityFindingSchema = z
  .object({
    kind: z.enum(["empty-dataset", "all-null", "high-null-rate", "constant", "candidate-key"]),
    severity: z.enum(["info", "warning", "error"]),
    column: columnNameSchema.nullable(),
  })
  .strict();

export const validationResultSchema = z
  .object({
    ruleIndex: z.number().int().nonnegative(),
    kind: validationRuleKindSchema,
    column: columnNameSchema,
    failedRows: z.number().int().nonnegative(),
    sampleRowNumbers: z.array(z.number().int().positive()).max(20),
  })
  .strict();

export const datasetQualityReportSchema = z
  .object({
    datasetId: datasetIdSchema,
    versionId: datasetIdSchema,
    rowCount: z.number().int().nonnegative(),
    score: z.number().int().min(0).max(100),
    columns: z.array(columnQualitySchema).min(1).max(500),
    findings: z.array(qualityFindingSchema).max(1_000),
    rules: validationRulesSchema,
    validation: z.array(validationResultSchema).max(100),
  })
  .strict();

export type ValidationRule = z.infer<typeof validationRuleSchema>;
export type DatasetValidationSaveInput = z.infer<typeof datasetValidationSaveInputSchema>;
export type ColumnQuality = z.infer<typeof columnQualitySchema>;
export type QualityFinding = z.infer<typeof qualityFindingSchema>;
export type ValidationResult = z.infer<typeof validationResultSchema>;
export type DatasetQualityReport = z.infer<typeof datasetQualityReportSchema>;

export function parseDatasetValidationSaveInput(value: unknown): DatasetValidationSaveInput {
  return datasetValidationSaveInputSchema.parse(value);
}

export function parseDatasetQualityReport(value: unknown): DatasetQualityReport {
  return datasetQualityReportSchema.parse(value);
}

File: services/data-core/internal/data/quality.go
package data

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
)

type qualityTarget struct {
	datasetID      string
	versionID      string
	tableName      string
	rowCount       int64
	columns        []ColumnProfile
	physicalByName map[string]string
}

func (service *Service) GetQualityReport(
	ctx context.Context,
	datasetID string,
) (DatasetQualityReport, error) {
	target, err := service.loadQualityTarget(ctx, datasetID)
	if err != nil {
		return DatasetQualityReport{}, err
	}
	rules, err := service.loadValidationRules(ctx, target)
	if err != nil {
		return DatasetQualityReport{}, err
	}
	validation := make([]ValidationResult, len(rules))
	for index, rule := range rules {
		result, err := service.evaluateValidationRule(ctx, target, index, rule)
		if err != nil {
			return DatasetQualityReport{}, err
		}
		validation[index] = result
	}
	columns, findings, score := deriveQuality(target, validation)
	return DatasetQualityReport{
		DatasetID:  target.datasetID,
		VersionID:  target.versionID,
		RowCount:   target.rowCount,
		Score:      score,
		Columns:    columns,
		Findings:   findings,
		Rules:      rules,
		Validation: validation,
	}, nil
}

func (service *Service) loadQualityTarget(ctx context.Context, datasetID string) (qualityTarget, error) {
	if !objectID.MatchString(datasetID) {
		return qualityTarget{}, errors.New("dataset id is invalid")
	}
	var target qualityTarget
	err := service.database.QueryRowContext(ctx, `
SELECT d.id, v.id, v.table_name, v.row_count
FROM datasets d
JOIN dataset_versions v ON v.id = d.current_version_id
WHERE d.id = ? AND v.status = 'ready'`, datasetID).Scan(
		&target.datasetID,
		&target.versionID,
		&target.tableName,
		&target.rowCount,
	)
	if errors.Is(err, sql.ErrNoRows) {
		return qualityTarget{}, errors.New("dataset not found")
	}
	if err != nil {
		return qualityTarget{}, fmt.Errorf("load quality target: %w", err)
	}
	if !internalTableName.MatchString(target.tableName) {
		return qualityTarget{}, errors.New("stored table name failed validation")
	}
	columns, physicalNames, err := service.loadColumns(ctx, target.versionID)
	if err != nil {
		return qualityTarget{}, err
	}
	target.columns = columns
	target.physicalByName = make(map[string]string, len(columns))
	for index, column := range columns {
		target.physicalByName[column.Name] = physicalNames[index]
	}
	return target, nil
}

func deriveQuality(
	target qualityTarget,
	validation []ValidationResult,
) ([]ColumnQuality, []QualityFinding, int) {
	columns := make([]ColumnQuality, len(target.columns))
	findings := make([]QualityFinding, 0)
	penalty := 0
	if target.rowCount == 0 {
		findings = append(findings, QualityFinding{Kind: "empty-dataset", Severity: "warning"})
		penalty += 25
	}
	for index, profile := range target.columns {
		nullRate := ratio(profile.NullCount, target.rowCount)
		distinctRate := ratio(profile.DistinctCount, target.rowCount-profile.NullCount)
		columns[index] = ColumnQuality{
			Name:          profile.Name,
			InferredType:  profile.InferredType,
			RowCount:      target.rowCount,
			NullCount:     profile.NullCount,
			NullRate:      nullRate,
			DistinctCount: profile.DistinctCount,
			DistinctRate:  distinctRate,
			MinValue:      profile.MinValue,
			MaxValue:      profile.MaxValue,
		}
		column := profile.Name
		switch {
		case target.rowCount > 0 && profile.NullCount == target.rowCount:
			findings = append(findings, QualityFinding{Kind: "all-null", Severity: "error", Column: &column})
			penalty += 25
		case nullRate >= 0.2:
			findings = append(findings, QualityFinding{Kind: "high-null-rate", Severity: "warning", Column: &column})
			penalty += 10
		}
		if target.rowCount-profile.NullCount > 1 && profile.DistinctCount == 1 {
			findings = append(findings, QualityFinding{Kind: "constant", Severity: "warning", Column: &column})
			penalty += 5
		}
		if target.rowCount > 1 && profile.NullCount == 0 && profile.DistinctCount == target.rowCount {
			findings = append(findings, QualityFinding{Kind: "candidate-key", Severity: "info", Column: &column})
		}
	}
	for _, result := range validation {
		if result.FailedRows > 0 {
			penalty += 15
		}
	}
	if penalty > 100 {
		penalty = 100
	}
	return columns, findings, 100 - penalty
}

func ratio(numerator, denominator int64) float64 {
	if denominator <= 0 {
		return 0
	}
	return float64(numerator) / float64(denominator)
}
